### PR TITLE
[IMP] hr_expense: Improved expense onboarding

### PR DESCRIPTION
**Pull request by hsal-odoo**
Improved onboarding, flow, and UX on desktop and mobile for expense. A more detailed overview could be found here https://app.excalidraw.com/l/65VNwvy7c4X/73HONs7UUCz

task-3919542

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/hr_expense/models/hr_expense.py [hunk 1]
```diff
@@ -642,18 +665,48 @@ def _get_empty_list_mail_alias(self):
...
+                'name': _("Expenses with a similar receipt to %s") % self.name,
+                'view_mode': 'tree,form',
+                'views': [
+                    [False, "tree"],
+                    [False, "form"],
+                ],
+                'res_model': 'hr.expense',
+                'domain': [('id', 'in', same_receipt_expense_ids.ids)],
+            }
+
```
Comment (hupo-odoo) [inline addons/hr_expense/models/hr_expense.py:709]:
```suggestion
        return self.same_receipt_expense_ids._get_records_action(
            name=_("Expenses with a similar receipt to %s") % self.name,
            domain=[('id', 'in', self.same_receipt_expense_ids.ids)],
        )
```
You can instead use this very useful new function :smile:

**Review by hupo-odoo — COMMENTED** [review]
Here is for now 2 comments, I'll keep reviewing tomorrow. You've done a great job! I know it's a hurry time so good job :smiley:

File: addons/hr_expense/views/hr_expense_views.xml [hunk 2]
```diff
@@ -568,36 +574,38 @@
...
-                                <field name="uom_id" class="w-25" groups="uom.group_uom" options="{'no_create': True}"/>
-                                <field name="uom_po_id" invisible="1"/>
-                                <label for="default_code"/>
-                                <div>
-                                    <field name="default_code" class="w-50"/>
-                                    <span class="d-inline-block">
-                                        <i class="text-muted">Use this reference as a subject prefix when submitting by email.</i>
+                                <field name="currency_id" invisible="1"/>
+                                <label for="standard_price"/>
+                                <div class="d-flex" groups="base.group_user">
```
Comment (hupo-odoo) [inline addons/hr_expense/views/hr_expense_views.xml:582]:
```suggestion
                                <div class="o_row" groups="base.group_user">
```
o_row seems seems better as it's the purpose of it

**Review by hupo-odoo — COMMENTED** [review]
Here are a few more comments :smile: Don't hesitate to tell if you don't agree with some comments of course

File: addons/hr_expense/static/src/scss/hr_expense.scss [hunk 3]
```diff
@@ -8,14 +8,42 @@
...
+    }
+
+    .o_view_pink_overlay {
+        border: 2px;
+        border-style: dashed;
+        border-radius: 10px;
+        border-color: pink;
+        background-color: #f2edf0;
+        width: 600px;
+        height: 300px;
```
Comment (mano-odoo) [inline addons/hr_expense/static/src/scss/hr_expense.scss:32]: Hello, we've been working this weekend with rugo to introduce a short animation when clicking on the "try a sample receipt" which should work in white mode and dark mode (task-3946192 & task-394615). While doing so I noticed that the pink overlay you introduced provides an unexpected visual result in darkmode.

image


I did a commit https://github.com/odoo/odoo/commit/a8de9376ee502b8e7e1a4237e2e7bd573cf57504 addressing this issue to merge with @rugo-odoo's work but since your PR is still under review it would be nice if you could adapt it here directly. This commit makes use of variables to reduce the styling while maintaining consistency between modules and making it work in darkmode. Feel free to add these change in your PR (the commit is from the branch `master-imp-expense_sample_receipt_animation-rugo-mano` but feel free to adapt it).

File: addons/hr_expense/static/src/components/expense_dashboard.js [hunk 4]
```diff
@@ -34,4 +35,15 @@ export class ExpenseDashboard extends Component {
         }
         return value;
     }
+
+    async applyFilter(filterName) {
+        const { actionId } = this.env.config;
+        const action = actionId ? await this.actionService.loadAction(actionId) : {};
```
[the remark this answers is not in the document]
Comment (hsal-odoo) [reply addons/hr_expense/static/src/components/expense_dashboard.js:41]: so the best u think is checking whether `this.env.config` contains `actionId`
and if not, I should throw some exception?

File: addons/hr_expense/static/src/views/list.js [hunk 5]
```diff
@@ -28,27 +34,28 @@ export class ExpenseListController extends ExpenseDocumentUpload(ListController)
 
     displaySubmit() {
         const records = this.model.root.selection;
-        return records.length && records.every(record => record.data.state === 'draft') && this.isExpenseSheet;
+        return records.length && records.every(record => record.data.state === 'draft') && this.isExpenseSheet && !this.onMobile;
```
[the remark this answers is not in the document]
Comment (hsal-odoo) [reply addons/hr_expense/static/src/views/list.js:37]: I need to hide these buttons on mobile using these methods, which are passed to `t-if` directives, in order to make them completely hidden on mobile. The approach followed before was to hide them using classes `d-none d-md-flex` which works fine, but has an issue. The issue is that the DOM acts as if these buttons exist, and it shows an arrow on mobile that when clicked should show the buttons, but when clicked, it shows nothing because the buttons are hidden. Laura complained about that, and that was my approach of fixing it.

File: addons/hr_expense/views/account_move_views.xml [hunk 6]
```diff
@@ -18,6 +18,10 @@
                             </div>
                     </button>
                 </xpath>
+
+                <xpath expr="//tree//field[@name='product_id']" position="attributes">
+                    <attribute name="domain">context.get('default_move_type') in ('out_invoice', 'out_refund', 'out_receipt') and [('sale_ok', '=', True)] or ['|', ('purchase_ok', '=', True), ('can_be_expensed', '=', True)]</attribute>
+                </xpath>
```
[the remark this answers is not in the document]
Comment (hsal-odoo) [reply addons/hr_expense/views/account_move_views.xml:24]: it is requested in the specs

[the remark this answers is not in the document]
Comment (hupo-odoo) [reply addons/hr_expense/views/account_move_views.xml:24]: Sorry I focused only on the Excalidraw and missed this info. My comment is wrong then. However, you should try to put several lines to not make one super long line, using ""
```
<attribute name="domain">
    context.get('default_move_type') in ('out_invoice', 'out_refund', 'out_receipt')
    and [('sale_ok', '=', True)]
    or ['|', ('purchase_ok', '=', True), ('can_be_expensed', '=', True)]
</attribute>
```

**Comment by mano-odoo** [comment]
Thanks a lot for applying the darkmode fix @hsal-odoo 🙏 

I can't review the PR because it's mostly functional changes but visually I noticed a few "issues" with the dropzone in the expenses app:

- The dropzone doesn't have the "hover" styling when a file is dragged on top of it (maybe intentional as I see in the kanban we still use this design? IMO triggering the same styling as the hover effect is a sufficient feedback, but if we keep the old design, then in list view the dropzone is currently broken as it's displayed under the dropzone)
image

- In kanban view the dropzone is not centered in the view
- In graph and pivot, the dropzone doesn't appear while the text is still there
- On mobile the design should probably be adapted because it's unlikely to drag and drop a file

Note that if this PR is really urgent the styling is still fixable after the freeze

File: addons/hr_expense/models/hr_expense.py [hunk 7]
```diff
@@ -435,6 +436,30 @@ def _compute_employee_id(self):
             for expense in self:
                 expense.employee_id = self.env.user.with_company(expense.company_id).employee_id
 
+    @api.depends('attachment_ids')
+    def _compute_same_receipt_expense_ids(self):
+        for expense in self:
+            expense.same_receipt_expense_ids = [Command.clear()]
```
Comment (JulienAlardot) [inline addons/hr_expense/models/hr_expense.py:442]:
```suggestion
       self.same_receipt_expense_ids = [Command.clear()]
```
? Should work the same :thinking: 
Not sure about how to improve this method but pretty sure it could :thinking:

**Review by JulienAlardot — COMMENTED** [review]
@epictete Do you have time for this one?

File: addons/hr_expense/models/hr_expense.py [hunk 8]
```diff
@@ -642,18 +667,31 @@ def _get_empty_list_mail_alias(self):
...
-                'send_string': _("Or send your receipts at"),
+                'send_string': _("Tip: try sending receipts by email"),
             }
         return ""
 
+    @api.model
+    def get_expense_attachments(self, expense_id):
+        expense = self.env['hr.expense'].search([('id', '=', expense_id)])
+        attachments = [attachment.image_src for attachment in expense.attachment_ids]
+        return attachments
```
Comment (william-andre) [inline addons/hr_expense/models/hr_expense.py:682]: Isn't this enough?
```suggestion
    def get_expense_attachments(self):
        return self.attachment_ids.mapped('image_src')
```

File: addons/hr_expense/models/hr_expense.py [hunk 9]
```diff
@@ -642,18 +667,31 @@ def _get_empty_list_mail_alias(self):
...
+
     # ----------------------------------------
     # Actions
     # ----------------------------------------
 
+    def action_show_same_receipt_expense_ids(self):
+        self.ensure_one()
+        return self.same_receipt_expense_ids._get_records_action(
+            name=_("Expenses with a similar receipt to %s") % self.name,
+            domain=[('id', 'in', self.same_receipt_expense_ids.ids)],
```
Comment (william-andre) [inline addons/hr_expense/models/hr_expense.py:692]: It's the default
```suggestion
```

File: addons/hr_expense/static/src/views/expense_line_widget.js [hunk 10]
```diff
@@ -26,10 +30,54 @@ export class ExpenseLinesListRenderer extends ListRenderer {
...
+    }
+
+    /** @override **/
+    async onGlobalClick(ev) {
+        const expenseName = this.props.record.data.name;
+        const attachments = await this.orm.call('hr.expense', 'get_expense_attachments', [this.props.record.resId]);
+        const files = attachments.map((attachment, index) => ({
+            isImage: true,
+            isViewable: true,
+            displayName: expenseName + `(${index + 1})`,
```
Comment (william-andre) [inline addons/hr_expense/static/src/views/expense_line_widget.js:47]: This might not behave nicely in all languages though?
Maybe using the translation function would be better
```suggestion
            displayName: expenseName + ` (${index + 1})`,
```

File: addons/hr_expense/static/src/views/list.js [hunk 11]
```diff
@@ -28,27 +34,28 @@ export class ExpenseListController extends ExpenseDocumentUpload(ListController)
...
     }
 
     displayCreateReport() {
-        const records = this.model.root.selection;
```
Comment (william-andre) [inline addons/hr_expense/static/src/views/list.js:35]: I don't know the JS modes so well, but I'm curious why this changes from `root.records` to `root.selection`, but not for the functions around this one

File: addons/hr_expense/views/account_move_views.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/hr_expense/views/account_move_views.xml:24]: This is really not modular though...
Isn't it valid to force expensable products to be purchasable on the model?

**Review by william-andre — COMMENTED** [review]
Just a quick technical scan, didn't read the spec or test the feature :see_no_evil:

File: addons/hr_expense/models/hr_expense.py (hunk 9 above)
Comment (hsal-odoo) [reply addons/hr_expense/models/hr_expense.py:692]: I don't get it? How would the default domain use `same_receipt_expense_ids` field?

File: addons/hr_expense/models/hr_expense.py (hunk 8 above)
Comment (hsal-odoo) [reply addons/hr_expense/models/hr_expense.py:682]: if I remove `api.model` how would I call it on self properly?

File: addons/hr_expense/static/src/views/expense_line_widget.js (hunk 10 above)
Comment (hsal-odoo) [reply addons/hr_expense/static/src/views/expense_line_widget.js:47]: Why not? and how should I use the translation function?

File: addons/hr_expense/static/src/views/list.js (hunk 11 above)
Comment (hsal-odoo) [reply addons/hr_expense/static/src/views/list.js:35]: There was some specs required related to this button, and that was one of them.

File: addons/hr_expense/models/hr_expense.py [hunk 12]
```diff
@@ -642,18 +667,31 @@ def _get_empty_list_mail_alias(self):
...
+        return attachments
+
     # ----------------------------------------
     # Actions
     # ----------------------------------------
 
+    def action_show_same_receipt_expense_ids(self):
+        self.ensure_one()
+        return self.same_receipt_expense_ids._get_records_action(
+            name=_("Expenses with a similar receipt to %s") % self.name,
```
Comment (JulienAlardot) [inline addons/hr_expense/models/hr_expense.py:691]:
```suggestion
            name=_("Expenses with a similar receipt to %(other_expense_name)s", other_expense_name=self.name),
```
Translation team will thank you

File: addons/hr_expense/models/hr_expense.py (hunk 9 above)
Comment (JulienAlardot) [reply addons/hr_expense/models/hr_expense.py:692]: > I don't get it? How would the default domain use `same_receipt_expense_ids` field?

Because you're calling it on self.same_receipt_expense_ids, so calling its self.ids == original self.same_receipt_expense_ids.ids

File: addons/hr_expense/static/src/views/list.js (hunk 11 above)
Comment (hsal-odoo) [reply addons/hr_expense/static/src/views/list.js:35]: image

Comment (william-andre) [reply addons/hr_expense/static/src/views/list.js:35]: But what is the difference between `this.model.root.records` and `this.model.root.selection`?

It looks like `action_show_expenses_to_submit` is still using `this.model.root.selection` (which makes sense...)


____

Please don't resolve conversations if you didn't fix anything. Or if I didn't say that I agreed with you (or resolved it myself)

Comment (hsal-odoo) [reply addons/hr_expense/static/src/views/list.js:35]: records -> all records in the list view 
selection -> selected records of the list view

The method `action_show_expenses_to_submit` indeed uses selection by default. But the method it calls (i.e., `get_expenses_to_submit` of the model `hr_expense`) returns all draft expenses if none were selected in the list view

Comment (william-andre) [reply addons/hr_expense/static/src/views/list.js:35]: What happens if you select only reported expenses?
Then it will submit a report for all the other unreported expenses?

Comment (JulienAlardot) [reply addons/hr_expense/static/src/views/list.js:35]: Creating 1 or 2 reports (if several payment_modes) with all unreported expenses of the user's employee if any, else raise

File: addons/hr_expense/static/src/views/expense_line_widget.js (hunk 10 above)
Comment (william-andre) [reply addons/hr_expense/static/src/views/expense_line_widget.js]: For right to left languages, they I believe they would expect the `(...)` on the other side (not 100% sure)

Comment (hsal-odoo) [reply addons/hr_expense/static/src/views/expense_line_widget.js]: I tested with Arabic, and it was added to the correct side 
image

File: addons/hr_expense/models/product_template.py [hunk 13]
```diff
@@ -17,6 +17,7 @@ def default_get(self, fields):
 
     can_be_expensed = fields.Boolean(string="Expenses", compute='_compute_can_be_expensed',
         store=True, readonly=False, help="Specify whether the product can be selected in an expense.")
+    purchase_ok = fields.Boolean('Purchase', default=True, compute='_compute_purchase_ok', store=True, readonly=False)
```
Comment (JulienVR) [inline addons/hr_expense/models/product_template.py:20]: `can_be_purchased` may be better ?

File: addons/hr_expense/models/product_template.py [hunk 14]
```diff
@@ -30,6 +31,12 @@ def _auto_init(self):
             )
         return super()._auto_init()
 
-    @api.depends('type')
+    @api.depends('type', 'purchase_ok')
     def _compute_can_be_expensed(self):
-        self.filtered(lambda p: p.type not in ['consu', 'service']).update({'can_be_expensed': False})
+        self.filtered(lambda p: p.type not in ['consu', 'service'] or not p.purchase_ok).update({'can_be_expensed': False})
```
Comment (JulienVR) [inline addons/hr_expense/models/product_template.py:36]: `self.filtered(lambda p: p.type not in ['consu', 'service'] or not p.purchase_ok).can_be_expensed = False`

File: addons/hr_expense/models/product_template.py [hunk 15]
```diff
@@ -30,6 +31,12 @@ def _auto_init(self):
...
+
+    @api.depends('can_be_expensed')
+    def _compute_purchase_ok(self):
+        for record in self:
+            if record.can_be_expensed:
+                record.purchase_ok = True
```
Comment (JulienVR) [inline addons/hr_expense/models/product_template.py:42]: `record.purchase_ok = record.can_be_expensed`

**Review by JulienVR — COMMENTED** [review]
First small batch, I'm still reviewing :)

File: addons/hr_expense/models/product_template.py (hunk 13 above)
Comment (hsal-odoo) [reply addons/hr_expense/models/product_template.py:20]: this is the name of the original field in the parent model

File: addons/hr_expense/models/product_template.py (hunk 14 above)
Comment (hsal-odoo) [reply addons/hr_expense/models/product_template.py:36]: why is this diff needed?

File: addons/hr_expense/models/product_template.py (hunk 13 above)
Comment (JulienVR) [reply addons/hr_expense/models/product_template.py:20]: oooh ok, mb

File: addons/hr_expense/models/product_template.py (hunk 14 above)
Comment (JulienVR) [reply addons/hr_expense/models/product_template.py:36]: doing an update with a dictionary containing only one element is useless, but yeah it's not your fault. Feel free to not do it :man_shrugging:

File: addons/hr_expense/models/product_template.py (hunk 15 above)
Comment (JulienVR) [reply addons/hr_expense/models/product_template.py:42]: but the default for a boolean is False if we don't specify anything, so `record.purchase_ok = record.can_be_expensed` is equivalent to what you have now, no ?

Comment (hsal-odoo) [reply addons/hr_expense/models/product_template.py:42]: yeah but what if `can_be_expensed` is set to False? This will set `purchase_ok` to False as well , and as far as I understand, this is not the needed behavior.

File: addons/hr_expense/views/hr_expense_views.xml [hunk 16]
```diff
@@ -568,36 +574,38 @@
...
+                                            using this product won't be able to change the amount of the expense,
+                                            only the quantity. Use a cost different than 0 for expense categories funded by
+                                            the company at fixed cost like allowances for mileage, per diem, accommodation
+                                            or meal."/>
+                                    <span class="d-flex w-50 ms-2" invisible="standard_price == 0">
+                                        per <field name="uom_id" class="ms-2" options="{'no_create': True}"/>
                                     </span>
                                 </div>
-                                <field name="categ_id" class="w-50"/>
+                                <field name="uom_po_id" invisible="1"/>
```
Comment (JulienVR) [inline addons/hr_expense/views/hr_expense_views.xml:596]: is it still needed ?

Comment (hsal-odoo) [reply addons/hr_expense/views/hr_expense_views.xml:596]: not in the current code, but maybe removing it would break some inheritance?

File: addons/hr_expense/models/product_template.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/hr_expense/models/product_template.py:20]: this should never be done in inheriting modules in order to keep Odoo's modularity

[the remark this answers is not in the document]
Comment (JulienVR) [reply addons/hr_expense/models/product_template.py:20]: mmh yeah that's right, we can still remove this in https://github.com/odoo/odoo/pull/167161

[the remark this answers is not in the document]
Comment (JulienVR) [reply addons/hr_expense/models/product_template.py:20]: @hsal-odoo :)
